common/Scripts/ModifyWindowsOptionalFeatures.cs

```diff
@@ -0,0 +1,227 @@
+            }
+            catch (Exception ex)
+            {
+                // This is most likely a case where the user cancelled the UAC prompt.
+                if (ex is System.ComponentModel.Win32Exception win32Exception)
+                {
+                    if (win32Exception.NativeErrorCode == 1223)
+                    {
+                        log?.Information(ex, "UAC was cancelled by the user.");
+                        exitCode = ExitCode.Cancelled;
+                    }
+                }
+                else
+                {
+                    log?.Error(ex, "Script failed");
+                    exitCode = ExitCode.Failure;
+                }
+            }
+        },
+            cancellationToken);
+
+        stopwatch.Stop();
+
+        ModifyWindowsOptionalFeaturesEvent.Log(
+            featuresString,
+            exitCode,
+            stopwatch.ElapsedMilliseconds);
+
+        return exitCode;
+    }
+
+    public enum ExitCode
+    {
+        Success = 0,
+        NoChange = 1,
+        Failure = 2,
+        Cancelled = 3,
+    }
+
+    private static ExitCode FromExitCode(int exitCode)
+    {
+        return exitCode switch
+        {
+            0 => ExitCode.Success,
+            1 => ExitCode.NoChange,
+            2 => ExitCode.Failure,
+            _ => ExitCode.Cancelled,
+        };
+    }
+
+    /// <summary>
+    /// PowerShell script for modifying Windows optional features.
+    ///
+    /// This script takes a string argument representing feature names and their desired states (enabled or disabled).
+    /// It parses this string into a dictionary, iterates over each feature, and performs the necessary enable or disable operation based on the desired state.
+    ///
+    /// The script defines the following possible exit statuses:
+    /// - OperationSucceeded (0): All operations (enable or disable) succeeded and no restart is needed.
+    /// - OperationSkipped (1): No operations were performed because the current state of all features matched the desired state.
+    /// - OperationFailed (2): At least one operation failed.
+    ///
+    /// Only features present in "validFeatures" are considered valid. If an invalid feature name is encountered, the script exits with OperationFailed.
+    /// This list should be kept consistent with the list of features in the WindowsOptionalFeatureNames class.
+    ///
+    /// </summary>
+    private const string Script =
+@"
+enum OperationStatus
+{
+    OperationSucceeded = 0
+    OperationSkipped = 1
+    OperationFailed = 2
+}
+
+$validFeatures = @(
+    'Containers',
+    'HostGuardian',
+    'Microsoft-Hyper-V-All',
+    'Microsoft-Hyper-V-Tools-All',
+    'Microsoft-Hyper-V',
+    'VirtualMachinePlatform',
+    'HypervisorPlatform',
+    'Containers-DisposableClientVM',
+    'Microsoft-Windows-Subsystem-Linux'
+)
+
+function ModifyFeatures($featuresString)
+{
+    $features = ConvertFrom-StringData $featuresString

```

Comment:
NIT of a NIT: `ConvertFrom-StringData` looks for new lines.  If this changes, the code that makes the arguments breaks.  If this changes the associated code needs to change as well.

Reply:
That's one reason I have everything contained in this one helper class. It's simple enough, this should be self explanatory for now.